networkx/algorithms/centrality/laplacian.py

```diff
@@ -0,0 +1,145 @@
+
+    Parameters
+    ----------
+
+    G : graph
+        A networkx graph
+
+    normalized : bool (default = True)
+        If True the centrality score is scaled so the sum over all nodes is 1.
+        If False the centrality score for each node is the drop in Laplacian
+        energy when that node is removed.
+
+    nodelist : list, optional (default = None)
+        The rows and columns are ordered according to the nodes in nodelist. If nodelist is None, then the ordering is produced by G.nodes().
+
+    weight: string or None, optional (default=`weight`)
+        Optional parameter for the Laplacian matrix calculation. The edge data key used to compute each value in the matrix. If None, then each edge has weight 1.
+
+    walk_type : string or None, optional (default=None)
+        Optional parameter for the Laplacian matrix calculation. If None, P is selected depending on the properties of the graph. Otherwise is one of `random`, `lazy`, or `pagerank`.
+
+    alpha : real (default = 0.95)
+        Optional parameter for the Laplacian matrix calculation. (1 - alpha) is the teleportation probability used with pagerank.
+
+    Returns
+    -------
+    nodes : dictionary
+       Dictionary of nodes with Laplacian centrality as the value.
+
+    Examples
+    --------
+    >>> G = nx.Graph()
+    >>> edges = [(0, 1, 4), (0, 2, 2), (2, 1, 1), (1, 3, 2), (1, 4, 2), (4, 5, 1)]
+    >>> G.add_weighted_edges_from(edges)
+    >>> sorted((v, f"{c:0.2f}") for v, c in laplacian_centrality(G).items())
+    [(0, '0.70'), (1, '0.90'), (2, '0.28'), (3, '0.22'), (4, '0.26'), (5, '0.04')]
+
+    Notes
+    -----
+    The algorithm is implemented based on [1] with an extension to directed graphs
+    using the `nx.directed_laplacian_matrix` function.
+
+    Raises
+    ------
+    NetworkXPointlessConcept
+        If the graph `G` is the null graph.
+
+    References
+    ----------
+    .. [1] Qi, X., Fuller, E., Wu, Q., Wu, Y., and Zhang, C.-Q. (2012).
+    Laplacian centrality: A new centrality measure for weighted networks.
+    Information Sciences, 194:240-253.
+    https://math.wvu.edu/~cqzhang/Publication-files/my-paper/INS-2012-Laplacian-W.pdf
+    """
+    import numpy as np
+    import scipy as sp
+    import scipy.linalg  # call as sp.linalg
+    import scipy.sparse  # call as sp.sparse
+
+    def eigh_f(A):
+        return sp.linalg.eigh(A.toarray(), eigvals_only=True)
+
+    if len(G) == 0:
+        raise nx.NetworkXPointlessConcept(
+            "cannot compute centrality for the null graph"
+        )
+
+    if nodelist != None:
+        nodeset = set(G.nbunch_iter(nodelist))
+        if len(nodeset) != len(nodelist):
+            raise nx.NetworkXError(
+                "nodelist contains duplicate nodes or nodes not in G"
+            )
+        full_nodelist = nodelist + [n for n in G if n not in nodeset]
+    else:
+        nodeset = set(G.nbunch_iter(nodelist))
+        full_nodelist = nodeset
```

Comment:
I think using a set here may cause problems - the `nodelist` parameter description in the docstring explicitly mentions that it determines node ordering, but sets do not preserve order.

If node ordering actually isn't important, then I'd vote to update the docstring. Either way, I think the else branch could be simplified to `full_nodelist = list(G)`.

Reply:
In the code block for the `if` clause (nodelist provided) the nodeset should be fine because it is only used to check for duplicates and nodes in G. It does not impact the order of the resulting `full_nodeset`.

The `else` clause should definitely be simplified to `full_nodelist(G)`.   Thanks!

Reply:
The purpose of the `nodelist` variable is not only checking for duplicates and nodes in G, but the `laplacian_centrality` is calculated for the nodes in the `nodelist`. You can see it at the beginning of the for loop. If `nodelist` is None that means someone wants to calculate the centrality for all the nodes. 

As the definition of this metric is:

"""The Laplacian Centrality of a node `i` is measured by the drop in the Laplacian Energy after deleting node `i` from the graph."""

First I have to calculate the energy of the entire graph and then I can calculate the metric for the nodes in `nodelist` or for the entire graph (if `nodelist =None`)

Reply:
The case when nodelist is None the code does not need to check whether nodes are in the graph, or duplicate. It should just set nodelist and full_nodelist to list of all the nodes in the graph. It does not need to be a set. And there is a minor memory savings if it uses the same object for the nodelist and the full_nodelist. So, I think the code should be something like:
```python
nodelist = full_nodelist = list(G)
```
And then replace the only place where nodeset is used with `nodelist`.  Note that this will work when nodelist is not None as well.  The only reason we should be creating a set here is to identify duplicate nodes in a provided nodelist.